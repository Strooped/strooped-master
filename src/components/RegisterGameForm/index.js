import { Formik } from 'formik';
import React from 'react';
import PropTypes from 'prop-types';
import { createGameRoom } from '../../utils/api/gameRoomApi';
import Input from '../form/Input';
import SelectInput from '../form/SelectInput';

const isModeInModes = (modeId, modes) => !!modes
  .find(mode => mode.id === parseInt(modeId, 10));

const RegisterGameForm = ({ onRegistered, modes }) => {
  const submitGameRoom = (values) => {
    createGameRoom(values)
      .then(room => onRegistered(room))
      .catch(err => console.error('Failed to create game-room', err));
  };

  return <Formik
    initialValues={{
      mode: modes[0] ? modes[0].id : '',
      name: '',
    }}
    validate={(values) => {
      const errors = {};

      if (!values.name || values.name.length < 2) {
        errors.name = 'Please include a name, with minimum 3 characters';
      }

      if (!values.mode) {
        errors.mode = 'Please select which mode you wish to play';
      } else if (!isModeInModes(values.mode, modes)) {
        errors.mode = 'Unknown mode selected. Please select one of the listed modes';
      }

      return errors;
    }}
    onSubmit={submitGameRoom}
  >
    {({
      handleSubmit,
      values,
      handleBlur,
      handleChange,
    }) => <form onSubmit={handleSubmit}>
      <SelectInput
        name="mode"
        label="Game mode"
        value={values.mode}
        onBlur={handleBlur}
        onChange={handleChange}
        required
      >
        <option value="" disabled>No game selected</option>
        {modes.map(mode => <option key={`mode-${mode.id}`} value={mode.id}>{mode.title}</option>)}
      </SelectInput>

      <Input
        name="name"
        label="Name of game"
        value={values.name}
        onBlur={handleBlur}
        onChange={handleChange}
        placeholder="Super game"
        required
      />

      <div className="field is-grouped">
        <button className="button is-link is-fullwidth" type="submit">Create game</button>
      </div>
    </form>}
  </Formik>;
};

RegisterGameForm.propTypes = {
  onRegistered: PropTypes.func.isRequired,
  modes: PropTypes.arrayOf(PropTypes.object),
};

export default RegisterGameForm;
