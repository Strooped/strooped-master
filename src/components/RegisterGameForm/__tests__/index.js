import React from 'react';
import { render } from 'enzyme';
import RegisterGameForm from '../index';

describe('<RegisterGameForm/>', () => {
  // Used to silence annoying useLayoutEffect complaint from React
  jest.spyOn(console, 'error').mockImplementation();

  it('should render with game modes provided', () => {
    const wrapper = render(<RegisterGameForm
      modes={[
        {
          id: 1,
          title: 'Mode 1',
        },
        {
          id: 2,
          title: 'Mode 2',
        },
      ]}
      onRegistered={() => null}
    />);

    expect(wrapper).toMatchSnapshot();
  });
});
