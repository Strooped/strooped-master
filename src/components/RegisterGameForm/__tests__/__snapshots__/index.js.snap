// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegisterGameForm/> should render with game modes provided 1`] = `
<form>
  <div
    class="gamemodeselect"
  >
    <button
      class="gamemodepreview"
    >
      <h3
        class="subtitle is-4"
      >
        Mode 1
      </h3>
      <p />
    </button>
    <button
      class="gamemodepreview"
    >
      <h3
        class="subtitle is-4"
      >
        Mode 2
      </h3>
      <p />
    </button>
  </div>
  <div
    class="field is-grouped"
  >
    <button
      class="button is-link is-fullwidth"
      disabled=""
      type="submit"
    >
      Create game
    </button>
  </div>
</form>
`;
